### ISSUE TITLE
Type argument inference causes crash in AsSuperVisitor

### ISSUE THREAD

**Issue by kelloggm** [typetools/checker-framework#2199]
Consider the following Java program:

```
public class TypeParamCrash {
    static class StrangeConstructorTypeArgs<K, V> {
	public StrangeConstructorTypeArgs(Abstract<String, byte[]> abs) { }
    }

    static abstract class Abstract<KEY, VALUE> { }

    static class Concrete<K, V> extends Abstract<K, V> { }

    static StrangeConstructorTypeArgs getStrangeConstructorTypeArgs() {
	return new StrangeConstructorTypeArgs(new Concrete<>());
    }
}
```

`javac` compiles this program, but running any Checker Framework checker results in a crash on the call to `new Concrete<>()`(tested: Nullness Checker, Value Checker). Making the type arguments explicit (i.e. replacing `new Concrete<>()` with `new Concrete<String, byte[]>()`) eliminates the crash and permits the program to compile.

This might be a duplicate of the several open issues relating to `AsSuperVisitor`: #2190, #2195, #2196, #2198. I wanted to report this because I've encountered it in the wild.

Here is the output of `-AprintErrorStack`:
```
jsr308/checker-framework git:(master) ▶ checker/bin/javac -processor nullness TypeParamCrash.java
error: AsSuperVisitor: Unexpected combination: type: DECLARED superType: ARRAY.
  type: Object
  superType: byte []
  Compilation unit: TypeParamCrash.java
  Last visited tree at line 1 column 1:
  public class TypeParamCrash {
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:14)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.defaultAction(AbstractAtmComboVisitor.java:52)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visitDeclared_Array(AbstractAtmComboVisitor.java:123)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:303)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:57)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:85)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:67)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:112)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visit(AbstractQualifierPolymorphism.java:400)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visitDeclared_Declared(AbstractQualifierPolymorphism.java:459)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visitDeclared_Declared(AbstractQualifierPolymorphism.java:284)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:307)
  org.checkerframework.framework.type.visitor.EquivalentAtmComboScanner.scan(EquivalentAtmComboScanner.java:56)
  org.checkerframework.framework.type.visitor.EquivalentAtmComboScanner.visit(EquivalentAtmComboScanner.java:36)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visit(AbstractQualifierPolymorphism.java:402)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visit(AbstractQualifierPolymorphism.java:363)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.access$100(AbstractQualifierPolymorphism.java:284)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism.annotate(AbstractQualifierPolymorphism.java:164)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.constructorFromUse(GenericAnnotatedTypeFactory.java:1404)
  org.checkerframework.checker.nullness.KeyForAnnotatedTypeFactory.constructorFromUse(KeyForAnnotatedTypeFactory.java:79)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitNewClass(TypeFromExpressionVisitor.java:263)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitNewClass(TypeFromExpressionVisitor.java:44)
  com.sun.tools.javac.tree.JCTree$JCNewClass.accept(JCTree.java:1532)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:53)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:36)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromExpression(AnnotatedTypeFactory.java:1238)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:1016)
  org.checkerframework.framework.flow.CFAbstractTransfer.getValueFromFactory(CFAbstractTransfer.java:201)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitNode(CFAbstractTransfer.java:577)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitNode(CFAbstractTransfer.java:93)
  org.checkerframework.dataflow.cfg.node.AbstractNodeVisitor.visitObjectCreation(AbstractNodeVisitor.java:333)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitObjectCreation(CFAbstractTransfer.java:910)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitObjectCreation(CFAbstractTransfer.java:93)
  org.checkerframework.dataflow.cfg.node.ObjectCreationNode.accept(ObjectCreationNode.java:62)
  org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:408)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysisBlock(Analysis.java:234)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:187)
  org.checkerframework.framework.flow.CFAbstractAnalysis.performAnalysis(CFAbstractAnalysis.java:91)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1225)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1143)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1487)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:260)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:291)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:159)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:85)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:965)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:494)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:487)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
error: AsSuperVisitor: Unexpected combination: type: DECLARED superType: ARRAY.
  type: @Initialized @NonNull Object
  superType: @Initialized @NonNull byte @Initialized @NonNull []
  Compilation unit: TypeParamCrash.java
  Last visited tree at line 1 column 1:
  public class TypeParamCrash {
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:14)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.defaultAction(AbstractAtmComboVisitor.java:52)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visitDeclared_Array(AbstractAtmComboVisitor.java:123)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:303)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:57)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:85)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:67)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:112)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visit(AbstractQualifierPolymorphism.java:400)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visitDeclared_Declared(AbstractQualifierPolymorphism.java:459)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visitDeclared_Declared(AbstractQualifierPolymorphism.java:284)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:307)
  org.checkerframework.framework.type.visitor.EquivalentAtmComboScanner.scan(EquivalentAtmComboScanner.java:56)
  org.checkerframework.framework.type.visitor.EquivalentAtmComboScanner.visit(EquivalentAtmComboScanner.java:36)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visit(AbstractQualifierPolymorphism.java:402)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visit(AbstractQualifierPolymorphism.java:363)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.access$100(AbstractQualifierPolymorphism.java:284)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism.annotate(AbstractQualifierPolymorphism.java:164)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.constructorFromUse(GenericAnnotatedTypeFactory.java:1404)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitNewClass(TypeFromExpressionVisitor.java:263)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitNewClass(TypeFromExpressionVisitor.java:44)
  com.sun.tools.javac.tree.JCTree$JCNewClass.accept(JCTree.java:1532)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:53)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:36)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromExpression(AnnotatedTypeFactory.java:1238)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:1016)
  org.checkerframework.framework.flow.CFAbstractTransfer.getValueFromFactory(CFAbstractTransfer.java:201)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitNode(CFAbstractTransfer.java:577)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitNode(CFAbstractTransfer.java:93)
  org.checkerframework.dataflow.cfg.node.AbstractNodeVisitor.visitObjectCreation(AbstractNodeVisitor.java:333)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitObjectCreation(CFAbstractTransfer.java:910)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitObjectCreation(CFAbstractTransfer.java:93)
  org.checkerframework.dataflow.cfg.node.ObjectCreationNode.accept(ObjectCreationNode.java:62)
  org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:408)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysisBlock(Analysis.java:234)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:187)
  org.checkerframework.framework.flow.CFAbstractAnalysis.performAnalysis(CFAbstractAnalysis.java:91)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1225)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1143)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1487)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:260)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:291)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:159)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:85)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:965)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:494)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
Note: TypeParamCrash.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
2 errors
```

**Comment by wmdietl** [comment]
Thanks for the report! It turns out this wasn't a duplicate of the other issues.
I've incorporated a fix in #2193
A test case with expected type errors would be good, but #2193 should fix the crashes.
Can you confirm whether it fixes your original issues?

**Comment by kelloggm** [comment]
#2193 does fix it, thanks :)
